src/toxcore/dht_new/server.rs

```diff
@@ -0,0 +1,630 @@
+        Client::new(precomputed_key, pk, addr.clone())
+    }
+    /**
+    Function to handle incoming packets. If there is a response packet,
+    send back it to the peer.
+    */
+    pub fn handle_packet(&mut self, packet: DhtUdpPacket) -> IoFuture<()>
+    {
+        let pk = Server::get_pk(&packet.1.clone());
+        let client = self.create_client(&packet.0, pk);
+        match packet.1 {
+            DhtPacket::PingRequest(packet) => {
+                debug!("Received ping request");
+                self.handle_ping_req(client, packet)
+            },
+            DhtPacket::PingResponse(packet) => {
+                debug!("Received ping response");
+                self.handle_ping_resp(client, packet)
+            },
+            DhtPacket::NodesRequest(packet) => {
+                debug!("Received NodesRequest");
+                self.handle_nodes_req(client, packet)
+            },
+            DhtPacket::NodesResponse(packet) => {
+                debug!("Received NodesResponse");
+                self.handle_nodes_resp(client, packet)
+            },
+            DhtPacket::DhtRequest(dr) => {
+                // The packet kind of DhtRequest is in encrypted payload,
+                // so decrypting is needed first.
+                let payload = dr.get_payload(&self.sk)
+                    .map(|p| p)
+                    .map_err(|e| {
+                        error!("deserialize DhtRequest payload fail {:?}", e);
+                        e
+                    });
+                match payload {
+                    Ok(DhtRequestPayload::NatPingRequest(pl)) => {
+                        debug!("Received nat ping request");
+                        self.handle_nat_ping_req(client, dr, pl)
+                    },
+                    Ok(DhtRequestPayload::NatPingResponse(pl)) => {
+                        debug!("Received nat ping response");
+                        self.handle_nat_ping_resp(client, dr, pl)
+                    },
+                    p => {
+                        error!("received packet are not handled {:?}", p);
+                        Box::new( future::err(
+                            Error::new(ErrorKind::Other,
+                                "received packet are not handled"
+                        )))
+                    },
+                }
+            },
+            ref p => {
+                error!("received packet are not handled {:?}", p);
+                Box::new( future::err(
+                    Error::new(ErrorKind::Other,
+                        "received packet are not handled"
+                )))
+            }
+        }
+    }
+
+    /**
+    handle received PingRequest packet, then create PingResponse packet
+    and send back it to the peer.
+    */
+    fn handle_ping_req(&mut self, client: Client, request: PingRequest) -> IoFuture<()>
+    {
+        if let Ok(payload) = request.get_payload(&self.sk) {
+            let resp_payload = PingResponsePayload {
```

Comment:
Remove `From<PingRequestPayload> for PingResponsePayload` since you don't use it.